sdk/core/allocator/main.cc

```diff
@@ -863,22 +863,54 @@ __cheriot_minimum_stack(0x90) ssize_t
 	return cap->quota;
 }
 
-__cheriot_minimum_stack(0xd0) int heap_quarantine_empty()
+__cheriot_minimum_stack(0xd0) int heap_quarantine_flush(Timeout *timeout)
 {
 	STACK_CHECK(0xd0);
-	LockGuard g{lock};
-	while (gm->heapQuarantineSize > 0)
+
+	if (!check_timeout_pointer(timeout))
+	{
+		return -EINVAL;
+	}
+
+	if (LockGuard g{lock, timeout})
 	{
-		if (!gm->quarantine_dequeue())
+		while (gm->heapQuarantineSize > 0)
 		{
-			revoker.system_bg_revoker_kick();
+			auto epoch = revoker.system_epoch_get();
+			if (!gm->quarantine_dequeue())
```

Comment:
This probably wants to be repeated, like `while (gm->quarantine_dequeue()) { ; }`, possibly with a notion of time elapsing, because there may be lots of (that is, more than 4) objects for us to remove from quarantine before we need to kick and wait for any revocation passes to take place.  We should re-check `gm->heapQuarantineSize > 0` afterwards in case the entire quarantine was on the finished or (already-revoked) pending rings.

Reply:
I've refactored this a lot, what do you think of the new version?

Reply:
Running `while(gm->quarantine_dequeue()) { ; }` twice with a revocation in the middle will get _most_ of the quarantine most of the time (anything finished or already revoked prior to its invocation, and anything in quarantine prior to its kick of the revoker), it won't necessarily get things freed during its own operation.  That appears to be deliberate, and is probably fine semantics, but the documentation could be a little more clear, perhaps.